docs/faq/distributed_training.md

```diff
@@ -0,0 +1,286 @@
+# Distributed training
+MXNet supports distributed training enabling us to leverage multiple machines for faster training.
+In this document, we describe how it works, how to launch a distributed training job and
+some environment variables which provide more control.
+
+## Type of parallelism
+There are two ways in which we can distribute the workload of training a neural network across multiple devices (can be either GPU or CPU).
+The first way is *data parallelism*, which refers to the case where each device stores a complete copy of the model.
+Each device works with a different part of the dataset, and the devices collectively update a shared model.
+These devices can be located on a single machine or across multiple machines.
+In this document, we describe how to train a model with devices distributed across machines in a data parallel way.
+
+When models are so large that they don't fit into device memory, then a second way called *model parallelism* is useful.
+Here, different devices are assigned the task of learning different parts of the model.
+Currently, MXNet supports Model parallelism in a single machine only. Refer [Training with multiple GPUs using model parallelism](https://mxnet.incubator.apache.org/versions/master/how_to/model_parallel_lstm.html) for more on this.
+
+## How does distributed training work?
+The architecture of distributed training in MXNet is as follows:
+#### Types of processes
+MXNet has three types of processes which communicate with each other to accomplish training of a model.
+- Worker: A worker node actually performs training on a batch of training samples.
+Before processing each batch, the workers pull weights from servers.
+The workers also send gradients to the servers after each batch.
+Depending on the workload for training a model, it might not be a good idea to run multiple worker processes on the same machine.
+- Server: There can be multiple servers which store the model's parameters, and communicate with workers.
+A server may or may not be co-located with the worker processes.
+- Scheduler: There is only one scheduler.
+The role of the scheduler is to set up the cluster.
+This includes waiting for messages that each node has come up and which port the node is listening on.
+The scheduler then lets all processes know about every other node in the cluster, so that they can communicate with each other.
+
+#### KV Store
+MXNet provides a key-value store, which is a critical component used for multi-device and distributed training.
```

Comment:
I think both multi-machine and multi-device are distributed training. The wording here is a bit confusing. What do you think?

Reply:
What about?
```
#### Worker Communication via the KVStore API
MXNet provides a key-value store, which is a critical component used in distributed training. The communication of parameters across processors on a single machine, as well across multiple machines, is relayed through one or more servers with a key-value store for the parameters. MXNet's KVStore API is the interface for workers to communicate with the one or more servers holding the key-value store. {now start with...} Workers `push` gradients...
```
I'm left wondering though - what's the worker-to-device mapping (is an 8-gpu system one worker or 8)? Do the params get rolled up on a node with multiple gpus before going to the server? Or does each device count as a worker and talks to a server directly? Also, since there seems to be support for multiple servers, it's not clear if the workers talk to all of them or if there's something else going on...

Reply:
@eric-haibin-lin @pracheer Ya, multi-device seems appropriate

@aaronmarkham I've modified the paragraph as per your suggestion and added a line about what happens when there are multiple devices on a single machine. I've also moved up the section which describes the distribution of keys on servers